### PR TITLE
Versions API improvements (9763)

### LINKED ISSUE #9763: Performance: Slow response for the versions API call with large number of files or versions
Hi Dataverse Team,

I just wanted to share with you something that affects datasets with high number of versions or files. 

 There have been some previously related issues that were already solved/closed. However, I strongly believe that the problem may still happen. I will backup the issue with some interesting data.

The below problem is happening when retrieving `dataset versions information` through the native API, hitting the endpoint: `http://demo.dataverse.org/api/datasets/<dataset-version>/versions`

**Given** that I am a user with large many files and versions in a dataset
**When** I retrieve all the dataset versions
**Then** I would like to receive a fairly "fast" response
**So** the user experience is smooth

# Current behavior
When the dataset has a large number of files, and also a large number of versions, the response time increments dramatically. This can be seen in the below table

| dataset id  | # of versions  | # of files | response size (MB) | response time  | # lines in response  |
|:---:|:---:|:---:|:---:|:---:|:---:|
| 767863 | 4 | 1349 | 2.92 | 11.42s | 120k |
| 396086 | 82 | 36 | 1.66 | 10.83s | 80k |
| 774618 | 14 | 69 | 0.51 | 3.57s | 23k |
| 770972 | 2 | 60 | 0.08 | 1.14s | 3.3k |

One of our concerns here is the fact that the dataset with id `767863` already takes a long time only having 4 versions, which means that once it reaches for instance 10 versions, it may be easily taking more than 20 seconds to respond, and could potentially cause a timeout in Dataverse.

Additionally, Dataverse currently returns as part of the response to all the files and their metadata for each of the versions available. That causes a very large response payload that may be unnecessary.

**Note:** The number of files also seem to affect the speed at which a version is published

# Expected behavior
To have the ability to retrieve **dataset versions information** in an efficient way that does not impact massively the response time.

# Possible solution
- To return basic metadata about each of the dataset versions available without file information that could potentially be the real problem.

- To review whether there are possible parallelization improvements

It is likely that the user will not necessarily need all the information for each of the datasets, normally, they would click on the version they are interested in, where we could actually perform another request such as `http://demo.dataverse.org/api/datasets/<dataset-id>/versions/<version>`

Thanks a lot!

### PR DESCRIPTION
**Pull request by landreev**
**What this PR does / why we need it**:

(copy-and-pasting from the comments in the linked issue): 

I'm addressing the performance issues in the versions api via a combination of the following approaches: 
- (optional) pagination is being added to `/api/datasets/{id}/versions`
- a new flag `includeFiles` is added to both `/api/datasets/{id}/versions` and `/api/datasets/{id}/versions/{vid}` (true by default), providing an option to drop the file information from the output
- when files are requested to be included, the filemetadatas are looked up using the "left joint fetch" optimizations that retrieve the information from the extended table tree in a single query, instead of allowing EJB to perform 1 SELECT query on each 1:1 relation for each FileMetadata entry. On a version with N filemetadatas this saves 3*N individual queries. This is not free resources-wise by any means (costs more memory primarily), but still appears to show measurable improvement on datasets with large numbers of versions and/or files. 

The following real life datasets from IQSS prod. service are used in the sample tests below: 

- 5255036: 237 versions (!), ~2,400 files. Files are relatively sparsely spread between versions. 
- 4554342: 2 versions, 1 pub. 1 draft. ~25,000 files (all appear in both versions).
- 2710927: a control "reasonable" dataset, 100 files, 3 published versions. 

`/api/datasets/{id}/versions`:

| dataset id  | # of versions  | # of files | response size | v5.14 response time | new response time | new size w/out files | new time w/out files |
|:---:|:---:|:---:|:---:|:---:|---:|---:|---:|
| 5255036 | 237 | 2400 | 38MB | 1m30s | 48s | 2MB | 12s | 
| 4554342 | 2 | 25000 | 17MB | 2m16s | 12s | 2.5K | 1s |
| 2710927 | 3 | 100 | .2MB | 1s | <1s | 26K | <1s |

`/api/datasets/{id}/versions/{vid}`:

| dataset id  | version  | # of files in version | response size | v5.14 response time | new response time | new size w/out files | new time w/out files |
|:---:|:---:|:---:|:---:|:---:|---:|---:|---:|
| 5255036 | 87.1 | 12 | 24K | 1s | <1s | 12K | <1s | 
| 4554342 | draft | 25000 | 17MB | 2m17s | 12s | 4K | <1s |
| 2710927 | 3.0 | 100 | 78K | 1s | <1s | 9K | <1s |

Note: `/api/datasets/4554342/versions` and `/api/datasets/4554342/versions/draft` produce the same amount of output because the former api was called without authentication, so only one, published version was included. 

Note: tests above were run on the dedicated IQSS test system (not in prod., which is a beefier and faster system). 

**Which issue(s) this PR closes**:

Closes #9763

**Special notes for your reviewer**:

Please take an extra look at the framework for caching the adjusted "embargo publication date" that I added, replacing the real time lookup that used to be in `Dataset.getCitationDatet()`, and especially the flyway script added in the PR. 

**Suggestions on how to test this**:

The description above outlines the way I tested the branch during dev. 

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**:

### CODE REVIEW

File: doc/sphinx-guides/source/api/native-api.rst
Comment (pdurbin) [inline doc/sphinx-guides/source/api/native-api.rst:886]: This sounds like a backward incompatible change. Can you please add a release note snippet about this?

(I'm a little nervous about what integrations will break.)

File: src/main/java/edu/harvard/iq/dataverse/Dataset.java
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/Dataset.java:707]: Yes, good to remove commented out code.

File: src/main/java/edu/harvard/iq/dataverse/search/SearchIncludeFragment.java
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/search/SearchIncludeFragment.java:208]: Should lower to fine.

File: src/main/java/edu/harvard/iq/dataverse/DatasetVersionServiceBean.java
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/DatasetVersionServiceBean.java:178]: I recognize this from #9693. Does it matter which PR is merged first?

**Review by pdurbin — COMMENTED** [review]
Just some quick comments.

File: src/main/java/edu/harvard/iq/dataverse/DatasetVersionServiceBean.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/DatasetVersionServiceBean.java:178]: I would prefer #9693 to be merged first, and then resolve the conflict in this branch (I added a database lookup optimization on top of the code introduced in that PR).

File: src/main/java/edu/harvard/iq/dataverse/search/SearchIncludeFragment.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/search/SearchIncludeFragment.java:208]: Not sure why it was set to .info in my brach...

File: doc/sphinx-guides/source/api/native-api.rst
Comment (landreev) [reply doc/sphinx-guides/source/api/native-api.rst:886]: Good point, thank you. It IS actually backward-compatible, as submitted in the PR. (i.e., `includeFiles` defaults to `true` as implemented). But the guide fragment needs to be changed to reflect that.

Comment (landreev) [reply doc/sphinx-guides/source/api/native-api.rst]: I have also added a release note.

Comment (pdurbin) [reply doc/sphinx-guides/source/api/native-api.rst]: Thanks, I'm glad we're preserving backward compatibility.

File: src/main/java/edu/harvard/iq/dataverse/search/SearchIncludeFragment.java
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/search/SearchIncludeFragment.java]: Gone now, great. Thanks.

**Review by pdurbin — COMMENTED** [review]
As discussed at standup, @landreev sounds happy to add some tests so I'm returning this to him. Otherwise, I'd say it's ready for QA.

**Comment by landreev** [comment]
OK, will take a look and resolve.

**Comment by landreev** [comment]
Moving back in review.

**Review by pdurbin — COMMENTED** [review]
I didn't run the code but it looks good. Conflicts resolved. Tests added. Docs look good.

I did notice a sql script naming conflict so I bumped the version. And I merged the latest from develop.

I'm only holding off on moving this to QA because Jenkins tests are failing (the "fire off installer" step). I started a thread in Slack about it: https://iqss.slack.com/archives/C010LA04BCG/p1697206082190619

**Comment by pdurbin** [comment]
I just kicked this off to get a second opinion from our container-based API test runner: https://github.com/gdcc/api-test-runner/actions/runs/6509472756

**Review by pdurbin — APPROVED** [review]
Jenkins is still failing but GitHub Actions passed: https://github.com/gdcc/api-test-runner/actions/runs/6509472756

Approved.

**Comment by pdurbin** [comment]
I ran Jenkins again. It's now saying `[ERROR]   DatasetsIT.getVersionFiles:3485 1 expectation failed.", "JSON path data[1].label doesn't match.", "Expected: test_1.txt", "  Actual: test_3.txt"`

@landreev should we be worried about this? From https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9883/17/consoleFull

**Review by pdurbin — APPROVED** [review]
We're having some testing challenges within Jenkins at the moment...

- https://iqss.slack.com/archives/C010LA04BCG/p1697466092510429
- https://iqss.slack.com/archives/C010LA04BCG/p1697466281385699

... but I'm pretty sure they are unrelated to this PR. Moving to ready for QA.

**Pull request merged**
